Add divertor monoblock simulation from Shimada2024 (#147)
idaholab/TMAP8: doc/content/examples/divertor_monoblock/index.md

# Divertor Monoblock During Pulsed Operation

> TMAP8 is used to model tritium transport in a divertor monoblock to elucidate the effects of pulsed operation (up to fifty 1600 s plasma discharge and cool-down cycles) on the tritium in vessel inventory source term and ex-vessel release term (i.e., tritium retention and permeation) for safety analysis. This example reproduces the results presented in [!cite](Shimada2024114438).

Reviewer: ```suggestion
TMAP8 is used to model tritium transport in a divertor monoblock to elucidate the effects of pulsed operation (up to fifty 1600-second plasma discharge and cool-down cycles) on the tritium in-vessel inventory source term and ex-vessel release term (i.e., tritium retention and permeation) for safety analysis. This example reproduces the results presented in [!cite](Shimada2024114438).
```

Would prefer spelling out "second" here. Since we're spelling it out and using it as an adjective, we must add the hyphen for correctness.

To simulate tritium and thermal transport, we define two sets of PDEs.
First, the strong form of the mass conservation equation for solute (mobile) T atoms, $C_s$, is written as:

\begin{equation} \label{eq:concentration}
\frac{\partial C_s}{\partial t} + \nabla \cdot (-D \nabla C_s) + \alpha_t \frac{C_t^e}{N} C_s - \alpha_r C_t = 0.
\end{equation}

We use three sets of mass conservation equations to calculate the behaviors of solute T atoms in three different materials (i.e., W, Cu, CuCrZr).
Second, the strong form of the conservation of energy equation is written as:

\begin{equation} \label{eq:temperature}
\rho c_p \frac{\partial T}{\partial t} - \nabla \cdot (k_T \nabla T) = 0.
\end{equation}

[tab:variables] lists all the symbols and definitions used in [eq:concentration] and [eq:temperature].

!table id=tab:variables caption=Nomenclature of all the variables and physical parameters used in [eq:concentration] and [eq:temperature].
| Symbol | Variable or Physical Property | Unit |
| --- | --- | --- |
| $C_s$ | Concentration of solute (mobile) species | m$^{−3}$ |
| $C_t$ | Concentration of trapped species | m$^{−3}$ |
| $T$ | Temperature | K |

> | $D$ | Diffusivity of solute (mobile) species | m$^{2}$ /s |

Reviewer: ```suggestion
| $D$ | Diffusivity of solute (mobile) species | m$^{2}$ s$^{-1}$ |
```

Style consistency.

> | $K_s$ | Solubility of solute (mobile) species | m$^{−3}$/Pa$^{1/2}$ |

Reviewer: ```suggestion
| $K_s$ | Solubility of solute (mobile) species | m$^{−3}$ Pa$^{-1/2}$ |
```

Style consistency. Also, this physical parameter is not used in either of the equations mentioned in the caption and should be reorganized to be closer to where it is used. This is not the only one I see ($\alpha_{r0}$, $E_{dt}$) that mis-align with the caption.

Alternatively a section for "nomenclature" could be created near the top of the problem description so that it serves as a key for the equations and relationships in the document for those physically relevant quantities with units (as opposed to the purely FEM nomenclature).

| $C_{total}$ | Total concentration of species $C_{total}=C_s + C_t$ | m$^{−3}$ |
| $C_t^e$ | Concentration of empty trapping sites | m$^{−3}$ |
| $C_t^0$ | Concentration of trapping sites $C_t^e=C_t^0-C_t$ | m$^{−3}$ |
| $\alpha_t$ | Trapping rate coefficient, $2.75 \times 10^{11}$ | s$^{−1}$ |
| $\alpha_r$ | Release rate coefficient $\alpha_r=\alpha_{r0} \exp⁡((-E_{dt})/(k_b T))$ | s$^{−1}$ |
| $\alpha_{r0}$ | Pre-exponential factor, $8.40 \times 10^{12}$ | s$^{−1}$ |
| $E_{dt}$ | Detrapping energy | eV |
| $N$ | Atomic number density | m$^{−3}$ |

> | $\rho$ | Density | g/m$^{3}$ |
> | $c_p$ | Specific heat | J/kg/K |
> | $k_T$ | Thermal conductivity | W/m/K |

Reviewer: ```suggestion
| $\rho$ | Density | g m$^{-3}$ |
| $c_p$ | Specific heat | J kg$^{-1}$ K$^{-1}$ |
| $k_T$ | Thermal conductivity | W m$^{-1}$ K$^{-1}$ |
```

Style consistency.

\end{equation}

Then, to solve for the PDEs and physical phenomena, we can select appropriate kernels and boundary conditions (BCs) from MOOSE’s extensive library. The following three subsections describe each kernel, BC, and numerical method utilized in the present work.

!listing test/tests/divertor_monoblock/divertor_monoblock.i link=false block=Variables

### Kernels

For the mass conservation equation, we use `ADTimeDerivative` and `ADMatDiffusion` to solve the 1$^{st}$ and 3$^{rd}$ terms of [eq:concentration_weak].
The TMAP8 kernels `TrappingNodalKernel` and `ReleasingNodalKernel` solve the 4$^{th}$ and 5$^{th}$ terms of [eq:concentration_weak],
respectively and simulate the trapping/release behavior of hydrogen isotopes in/from trap sites.
For the conservation of energy equation, `SpecificHeatConductionTimeDerivative` and `HeatConduction` solve the 1st and 3rd terms of [eq:temperature_weak].
`ADTimeDerivative,` `ADMatDiffusion,` `SpecificHeatConductionTimeDerivative,` and `HeatConduction` are MOOSE kernels commonly
used to solve the time derivative, material diffusion in the material, and heat conduction equations in the material.

> !listing test/tests/divertor_monoblock/divertor_monoblock.i link=false block=Kernels

Author: I created https://github.com/idaholab/moose/issues/27820 and tagged you in it.

> !listing test/tests/divertor_monoblock/divertor_monoblock.i link=false block=Kernels

Author: In the meantime, I listed the kernel blocks before the auxKernel blocks in the input files.

Nodal kernels are used for non-diffusive variables (trapped species).

!listing test/tests/divertor_monoblock/divertor_monoblock.i link=false block=NodalKernels

### Boundary conditions and history

For the mass conservation equation, we use `FunctionNeumannBC` and `DirichletBC` to solve the 2$^{nd}$ term of [eq:concentration_weak].
`FunctionNeumannBC` is used to treat the time-dependent plasma exposure at the top (plasma-exposed) surface,
and `DirichletBC` is used to set the BC of the solute T atom concentration to zero at the inner CuCrZr tube (at $r = 6$ mm).
For the energy conservation equation, we use `FunctionNeumannBC` and `DirichletBC` to solve the 2$^{nd}$ term of [eq:temperature_weak].
`FunctionNeumannBC` is used to treat the time-dependent heat flux at the top (plasma-exposed) surface,
and `DirichletBC` is used to treat the temperature increase in the cooling tube.
`FunctionNeumannBC,` `FunctionDirichletBC,` and `DirichletBC` are MOOSE BCs commonly used to determine the BCs of the variables to be solved.

> We simulate a 1,600 s-cycle plasma discharge, with each cycle consisting of a 100 s plasma ramp-up,
> a 400 s steady-state plasma discharge, a 100 s plasma ramp-down, and a 1,000 s waiting phase.
> Up to 50 cycles are simulated to achieve the total 20,000 s steady-state plasma discharge.

Reviewer: ```suggestion
We simulate a 20,000-second plasma discharge, with each 1,600-second cycle consisting of a 100-second plasma ramp-up,
a 400-second steady-state plasma discharge, a 100-second plasma ramp-down, and a 1,000-second waiting phase.
Up to 50 cycles are simulated to achieve the total discharge.
```

> [fig:tritium_temperature_history] shows the integrated (solute, total and trapped) tritium concentration in the monoblock.

Reviewer: ```suggestion
[fig:tritium_temperature_history] shows the integrated (solute, total and trapped) tritium concentration profiles in the monoblock.
```

> It shows that implantation fluxes and temperatures varies linearly up to (from) their steady-state values from (up to) their initial values during ramp-up (ramp-down).

Reviewer: ```suggestion
It shows that implantation fluxes and temperatures vary linearly up to (from) their steady-state values from (up to) their initial values during ramp-up (ramp-down).
```

> During the steady-state plasma discharge, we set a heat flux of 10 MW/m$^{2}$ at the top of the 2D monoblock (at $y = 14.0$ mm)

Reviewer: ```suggestion
During the steady-state plasma discharge, we set a heat flux of 10 MW m$^{-2}$ at the top of the 2D monoblock (at $y = 14.0$ mm)
```

Style consistency.

and a cooling temperature of 552 K at the inner CuCrZr tube (at $r = 6.0$ mm).

> We assume 100% T plasma with a 5.0 $\times$ 10$^{23}$ m$^{-2}$$\cdot$s$^{-1}$ plasma particle flux

Reviewer: ```suggestion
We assume a 100% T plasma with a 5.0 $\times$ 10$^{23}$ m$^{-2}$$\cdot$s$^{-1}$ plasma particle flux
```

> (which is 50% of 1.0 $\times$ 10$^{24}$ DT m$^{-2}$$\cdot$s$^{-1}$ plasma particle flux)

Reviewer: ```suggestion
(which is 50% of 1.0 $\times$ 10$^{24}$ DT m$^{-2}$$\cdot$s$^{-1}$ plasma particle flux),
```

and only 0.1% of the incident plasma particle flux (5.0 $\times$ 10$^{20}$ m$^{-2}$$\cdot$s$^{-1}$) entered the first layer of mesh at $y = 14.0$ mm.

> We treated this plasma exposure by setting the flux BC (5.0 $\times$ 10$^{16}$ m$^{-3}$$\cdot$s$^{-1}$) of the solute T atom concentration in the first layer of mesh

Reviewer: Can you explain how you get this flux and unit? Also, the way this is presented in somewhat dense and hard to follow given the back-to-back flux quantities and scientific notation. I recommend an adjustment for readability.

Author: Hopefully the updated documentation clarifies things.

(1.0 $\times$ 10$^{–4}$ m) at $y = 14.0$ mm in order to avoid simulating the complex plasma implantation and recombination phenomena
via an extremely fine mesh size (∼10$^{–10}$ m) and a long computation time.
The solute T atom concentration is set to zero at the inner CuCrZr tube (at $r = 6.0$ mm).

!media examples/figures/divertor_monoblock_history.png
  id=fig:tritium_temperature_history
  caption=Temperature profiles (orange) and integrated tritium concentration profiles (blue) during two 1,600-second-cycle plasma discharges. This corresponds to Fig. 2 in Ref. [!cite](Shimada2024114438).
  style=display:block;margin-left:auto;margin-right:auto;width:60%

> !listing test/tests/divertor_monoblock/divertor_monoblock.i link=false block=BCs

Reviewer: Should the functions used be discussed as their own section since they are common in these BC objects? They are common for the material properties too.

Author: I added a section for the functions

### Material properties

> 

Reviewer: ```suggestion

!style halign=left
```

We use the tritium mass transport properties listed in [tab:material_properties_tritium]
to solve the mass conservation equation ([eq:concentration_weak]) for two variables:
$C_s$ and $C_t$ (i.e., solute and trapped T atom concentrations) in three materials
and the thermal properties (i.e., density, temperature-dependent specific heat, and thermal conductivity) listed in [tab:material_properties_thermal]
to solve the energy conservation equation ([eq:temperature_weak]) for one variable: the temperature $T$.
The diffusivity is defined as $D=D_0 \exp⁡(-E_D/k_B/T)$ and the solubility is defined as $K_s=K_{s,0} \exp⁡(-E_s/k_B/T)$.

!table id=tab:material_properties_tritium

>   caption=Tritium mass transport and trapping properties used in the W, Cu, and CuCrZr layers. NOTE: Two-component solubility in W kept the maximum solubility ratio between W and Cu to 104 at low temperature. The references are provided in [!cite](Shimada2024114438).

Reviewer: ```suggestion
  caption=Tritium mass transport and trapping properties used in the W, Cu, and CuCrZr layers. **NOTE: Two-component solubility in W kept the maximum solubility ratio between W and Cu to 104 at low temperature. The references are provided in [!cite](Shimada2024114438).
```

| Material | $D_0$ (m$^2$/s) | $E_D$ (eV) | $K_{s,0}$ (Pa$^{1/2}$) | $E_s$ (eV) | Detrapping energy: $E_{dt}$ (eV) | Trap density: $n_{trap}$ (at.fr.) |
| --- | --- | --- | --- | --- | --- | --- |
| W | 2.4$\times$10$^{-7}$ | 0.39 | 1.87$\times$10$^{24}$ | 1.04 | 0.85 | 1.0$\times$10$^{-4}$ |

> | W |   |   |  3.14$\times$10$^{20}$ | 0.57 |   |   |

Reviewer: ```suggestion
| W |   |   |  3.14$\times$10$^{20}$** | 0.57** |   |   |
```

| Cu | 6.6$\times$10$^{-7}$ | 0.39 | 3.14$\times$10$^{24}$ | 0.57 | 0.50 | 5.0$\times$10$^{-5}$ |
| CuCrZr | 3.9$\times$10$^{-7}$ | 0.42 | 4.28$\times$10$^{23}$ | 0.39 | 0.83 | 5.0$\times$10$^{-5}$ |

!table id=tab:material_properties_thermal
  caption=Thermal properties used in the W, Cu, and CuCrZr layers. The references are provided in [!cite](Shimada2024114438).

> | Material | Density: $\rho$ (g/m$^3$) | Specific heat: $c_p$ (J/kg/K) | Thermal conductivity: $k_T$ (W/m/K) |

Reviewer: ```suggestion
| Material | Density: $\rho$ (g m$^{-3}$) | Specific heat: $c_p$ (J kg$^{-1}$ K$^{-1}$) | Thermal conductivity: $k_T$ (W m$^{-1}$ K$^{-1}$) |
```
Style consistency.

| --- | --- | --- | --- |

> | W | 19,300 | 1.16$\times$10$^{2}$ +7.11$\times$10$^{-2}$ T –6.58$\times$10$^{-5}$ T2 +3.24$\times$10$^{-8}$ T3 –5.45$\times$10$^{-12}$ T4 (293 < T(K)<2500) | 2.41$\times$10$^{2}$ –2.90$\times$10$^{-1}$ T + 2.54$\times$10$^{-4}$ T2 –1.03$\times$10$^{-7}$ T3 +1.52$\times$10$^{-11}$ T4 (293 < T(K) <2500) |

Reviewer: ```suggestion
| W | 19,300 | 1.16$\times$10$^{2}$ +7.11$\times$10$^{-2}$ T –6.58$\times$10$^{-5}$ T$^{2}$ +3.24$\times$10$^{-8}$ T$^{3}$ –5.45$\times$10$^{-12}$ T$^{4}$ (293 < T (K) < 2500) | 2.41$\times$10$^{2}$ –2.90$\times$10$^{-1}$ T + 2.54$\times$10$^{-4}$ T$^{2}$ –1.03$\times$10$^{-7}$ T$^{3}$ +1.52$\times$10$^{-11}$ T$^{4}$ (293 < T (K) < 2500) |
```

> | Cu | 8,960 | 4.21$\times$10$^{2}$ –6.85$\times$10$^{-2}$ T (293 < T(K) <873) | 3.16$\times$10$^{2}$ +3.18$\times$10$^{-1}$ T –3.49$\times$10$^{-4}$ T2 +1.66$\times$10$^{-7}$ T3 (293 < T(K) <873) |

Reviewer: ```suggestion
| Cu | 8,960 | 4.21$\times$10$^{2}$ –6.85$\times$10$^{-2}$ T (293 < T (K) < 873) | 3.16$\times$10$^{2}$ +3.18$\times$10$^{-1}$ T –3.49$\times$10$^{-4}$ T$^{2}$ +1.66$\times$10$^{-7}$ T$^{3}$ (293 < T (K) < 873) |
```

> | CuCrZr | 8,900 | 390 | 3.87$\times$10$^{2}$ –1.28$\times$10$^{-1}$ T (293 < T(K)<927) |

Reviewer: ```suggestion
| CuCrZr | 8,900 | 390 | 3.87$\times$10$^{2}$ –1.28$\times$10$^{-1}$ T (293 < T (K) < 927) | 
```

!listing test/tests/divertor_monoblock/divertor_monoblock.i link=false block=Materials

# Interface Kernels

TMAP8 assumes equilibrium between the chemical potentials of the diffusing species similar to how
TMAP4 and TMAP7 treated the diffusing species across two different materials with different chemical potentials.

> `SolubilityRatioMaterial` is used to treat solute concentration differences stemming from a difference in T solubilities across the interface.

Reviewer: ```suggestion
[SolubilityRatioMaterial.md] is used to treat solute concentration differences stemming from a difference in T solubilities across the interface.
```

The solubility ratio jump is calculated via the following:
\begin{equation} \label{eq:solubility}
J = \frac{C_{s,1}}{K_{s,1}} - \frac{C_{s,2}}{K_{s,2}}.
\end{equation}

> `ADPenaltyInterfaceDiffusion` is used to conserve the particle flux at this interface between two different solubilities.

Reviewer: ```suggestion
[ADPenaltyInterfaceDiffusion.md] is used to conserve the particle flux at this interface between two different solubilities.
```

The extremely low tritium solubility in W at low temperature leads to an extremely low solute concentration in W,
creating a large solute concentration difference between W and Cu.
Two-component solubility in W is used to keep the maximum solubility ratio between W and Cu to 104 at low temperature
to avoid the convergence issue associated with calculating two significantly different solute concentration in W and Cu.

!listing test/tests/divertor_monoblock/divertor_monoblock.i link=false block=InterfaceKernels

### Numerical method

> We use a standard preconditioner; the “single matrix preconditioner”.

Reviewer: ```suggestion
We use a standard preconditioner: the [“single matrix preconditioner”](SingleMatrixPreconditioner.md).
```

> The Newton method is used to model `transient` tritium and thermal transport in a 2D monoblock.

Reviewer: ```suggestion
The Newton method is used to model the [transient](Transient.md) tritium and thermal transport in a 2D monoblock.
```

It is important to note that MOOSE is equipped with the built-in Message Passing Interface (MPI) protocol,

> as tritium and thermal transport analysis of fifty 1,600 s-cycle plasma discharges in the 2D monoblock is

Reviewer: ```suggestion
as tritium and thermal transport analysis of fifty 1,600-second cycle plasma discharges in the 2D monoblock is
```

> performed in under 2 h using a single device/computer (3.5 GHz Apple M2 Pro, 10-Core CPU/16-Core GPU) with this MPI feature.

Reviewer: ```suggestion
performed in under 2 hours using a single device/computer (3.5 GHz Apple M2 Pro, 10-Core CPU/16-Core GPU) with this MPI feature.
```

!listing test/tests/divertor_monoblock/divertor_monoblock.i link=false block=Executioner

## Results

The simulation results from this example are shown in [fig:results2D_a] and [fig:results2D_b].
For more results, information, and discussion about the results for this example case and

> their significance, the reader is referred to Ref. [!cite](Shimada2024114438).

Reviewer: ```suggestion
their significance, the reader is referred to [!cite](Shimada2024114438).
```

!media examples/figures/divertor_monoblock_results_2D_a.png
  id=fig:results2D_a

>   caption=Tritium concentration profile in W (left), Cu (center), and CuCrZr (right) after ten 1,600 s cycles. This corresponds to Fig. 4A in Ref. [!cite](Shimada2024114438).

Reviewer: ```suggestion
  caption=Tritium concentration profile in W (left), Cu (center), and CuCrZr (right) after ten 1,600-second cycles. This corresponds to Fig. 4A in [!cite](Shimada2024114438).
```

  style=display:block;margin-left:auto;margin-right:auto;width:40%

!media examples/figures/divertor_monoblock_results_2D_b.png
  id=fig:results2D_b

>   caption=Tritium concentration profile in W (left), Cu (center), and CuCrZr (right) after fifty 1,600 s cycles. This corresponds to Fig. 4B in Ref. [!cite](Shimada2024114438).

Reviewer: ```suggestion
  caption=Tritium concentration profile in W (left), Cu (center), and CuCrZr (right) after fifty 1,600-second cycles. This corresponds to Fig. 4B in [!cite](Shimada2024114438).
```

  style=display:block;margin-left:auto;margin-right:auto;width:40%

!alert warning title=The exodus file in `gold` is a smaller version of the output

> The input file [/divertor_monoblock.i] returns the outputs that were used in Ref. [!cite](Shimada2024114438). However, a slightly modified version of this input is run in [/divertor_monoblock/tests] as part of TMAP8's [Software Quality Assurance](sqa/index.md) process: It simulates only one pulse cycle, has a coarser mesh, and outputs the results less regularly to limit the file size. As a result, the exodus file on `gold` is a smaller version of the output generated when running the full input file.

Reviewer: ```suggestion
The input file [/divertor_monoblock.i] returns the outputs that were used in [!cite](Shimada2024114438). However, a slightly modified version of this input is run in [/divertor_monoblock/tests] as part of TMAP8's [Software Quality Assurance](sqa/index.md) process: It simulates only one pulse cycle, has a coarser mesh, and outputs the results less regularly to limit the file size. As a result, the exodus file in the test `gold` directory is a smaller version of the output generated when running the full input file.
```

## Complete input file

> Below is the complete input file, which can be run with around 4 processors. Note that this input file has not been optimized for computational costs.

Reviewer: ```suggestion
Below is the complete input file, which can be run reliably with approximately 4 processor cores. Note that this input file has not been optimized for computational costs.
```

`FunctionNeumannBC`, `FunctionDirichletBC`, `DirichletBC`, and `FunctionDirichletBC` are MOOSE objects commonly used to represent the BCs of the variables to be solved.

We simulate a 20,000-second plasma discharge, with each 1,600-second cycle consisting of a 100-second plasma ramp-up,
a 400-second steady-state plasma discharge, a 100-second plasma ramp-down, and a 1,000-second waiting phase.
Up to 50 cycles are simulated to achieve the total discharge.
[fig:tritium_temperature_history] shows the integrated (solute, total and trapped) tritium concentration profiles in the monoblock.
It shows that implantation fluxes and temperatures vary linearly up to (from) their steady-state values from (up to) their initial values during ramp-up (ramp-down).

!media examples/figures/divertor_monoblock_history.png
  id=fig:tritium_temperature_history
  caption=Temperature profiles (orange) and integrated tritium concentration profiles (blue) during two 1,600-second-cycle plasma discharges. This corresponds to Fig. 2 in [!cite](Shimada2024114438).
  style=display:block;margin-left:auto;margin-right:auto;width:60%

> During the steady-state plasma discharge, we set a heat flux of 10 MW m$^{-2}$ at the top of the 2D monoblock (at $y = 14.0$ mm)

Author: I should have caught it, though. Good catch!

### Functions

!style halign=left

In this example, [Functions](Functions/index.md) are used to define the time-dependent tritium
and heat flux at the exposed surface, as well as the coolant temperature.
Functions can also be spatially-dependent.
The functions are then provided to BCs, as described above.

!listing test/tests/divertor_monoblock/divertor_monoblock.i link=false block=Functions

### Material properties

!style halign=left
We use the tritium mass transport properties listed in [tab:material_properties_tritium]
to solve the mass conservation equation ([eq:concentration_weak]) for two variables:
$C_s$ and $C_t$ (i.e., solute and trapped T atom concentrations) in three materials
and the thermal properties (i.e., density, temperature-dependent specific heat, and thermal conductivity) listed in [tab:material_properties_thermal]
to solve the energy conservation equation ([eq:temperature_weak]) for one variable: the temperature $T$.
The diffusivity is defined as $D=D_0 \exp⁡(-E_D/k_B/T)$ and the solubility is defined as $K_s=K_{s,0} \exp⁡(-E_s/k_B/T)$.

!table id=tab:material_properties_tritium

>   caption=Tritium mass transport and trapping properties used in the W, Cu, and CuCrZr layers. **NOTE: Two-component solubility in W kept the maximum solubility ratio between W and Cu to 104 at low temperature. The references are provided in [!cite](Shimada2024114438).

Author: Great catch!
Fixed now.
